Add --flush-cache for ad-hoc commands

anvitpusalkar opened the pull request:
##### SUMMARY
Added --flush-cache for ad-hoc commands

Fixes: #83749

##### ISSUE TYPE

- Feature Pull Request

##### ADDITIONAL INFORMATION

<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

Modified options_helpers.py to include --flush-cache in add_inventory_options.
Modified console.py to include --flush-cache as CLI option.

Tested ansible ad-hoc command so that it shows --flush-cache as a CLI option and it's usage.

Created ansible.cfg to create facts_cache:
[defaults]
fact_caching = jsonfile
fact_caching_connection = <path-to-cache_file>

Run command to verify cache_file is created:
ansible localhost -m setup

Run command with flush_cache option:
ansible --flush-cache localhost -m setup

<!--- Paste verbatim command output below, e.g. before and after your change -->

Command with --flush-cache is successful.
```paste below

```

anvitpusalkar committed:
Issue:83749 add --flush-cache for ad-hoc commands

anvitpusalkar committed:
Issue:83749 add --flush-cache for ad-hoc commands

Akasurde commented on lib/ansible/cli/arguments/option_helpers.py:
```diff
@@ -297,6 +297,8 @@ def add_inventory_options(parser):
                         help='outputs a list of matching hosts; does not execute anything else')
     parser.add_argument('-l', '--limit', default=C.DEFAULT_SUBSET, dest='subset',
                         help='further limit selected hosts to an additional pattern')
+    parser.add_argument('--flush-cache', dest='flush_cache', action='store_true',
```
Please create a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment). See [this fragment](https://github.com/ansible/ansible/blob/c2f528b7686c9573aa8493357219e0bbe4416dd3/changelogs/fragments/reboot-change-default-boot-command.yaml) as an example.

anvitpusalkar replied:
Added changelog fragment file 84149-add-flush-cache-for-adhoc-commands.yml

anvitpusalkar committed:
add --flush-cache option for the ansible command

Akasurde commented:
> CI (Remote macOS 14.3 - 4) test seems to fail regularly, is this a known issue?

It seems intermittent, restarted failed CI run.

s-hertel commented on lib/ansible/cli/console.py:
```diff
@@ -558,6 +558,10 @@ def run(self):
 
         hosts = self.get_host_list(self.inventory, context.CLIARGS['subset'], self.pattern)
 
+        # flush fact cache if requested
+        if context.CLIARGS['flush_cache']:
+            self._flush_cache(self.inventory, self.variable_manager)
+
```
I think it might be better to do this in the CLI base class `_play_prereqs` method instead (then you can also delete the pattern from playbook.py).

Right now adhoc does flush the cache for inventory plugins, but not the fact cache (which would be fixed by this logic happening in the base class instead). After the setup task, try running `ansible localhost --flush-cache -m debug -a "msg={{ ansible_facts }}"` to see that the fact cache isn't getting cleared yet. It would also be great to add tests for this (maybe in test/integration/targets/adhoc).

anvitpusalkar replied:
Moving this logic to the CLI base class seems to work. After running setup task, the fact cache is getting cleared as seen below:
```
ansible localhost --flush-cache -m debug -a "msg={{ ansible_facts }}"

<output>
localhost | SUCCESS => {
    "msg": {}
}
```

However  _flush_cache needs to handle "localhost" separately since the hosts_list is empty. Testing further...

s-hertel replied:
Good point. It's probably simpler to fix in adhoc after the host list is retrieved.

anvitpusalkar replied:
Would it be okay to add something like below to always handle `localhost` whenever `--flush-cache` is used.

```
@staticmethod
    def _flush_cache(inventory, variable_manager):
        **variable_manager.clear_facts('localhost')**
        for host in inventory.list_hosts():
            hostname = host.get_name()
            variable_manager.clear_facts(hostname)
```

s-hertel replied:
Yes, adding `variable_manager.clear_facts('localhost')` seems like a good solution.

s-hertel commented on lib/ansible/cli/arguments/option_helpers.py:
```diff
@@ -297,6 +297,8 @@ def add_inventory_options(parser):
+                        help="clear the fact cache for every host in inventory")
 
 
 def add_meta_options(parser):
```
I think you could also delete the duplicate option from `add_meta_options`.

anvitpusalkar commented:
Tried running the ansible command using an inventory file:
```
[servers]
server1 ansible_host=<ip-addr> ansible_ssh_private_key_file=<path>
```

**Run setup task:**
`ansible servers -i ../ansible-test/inventory -m setup`

Facts cache is created.

**Test without --flush-cache:**

```
ansible servers -i ../ansible-test/inventory --flush-cache -m debug -a "msg={{ ansible_facts }}"

<output>
server1 | SUCCESS => {
    "msg": {
        "all_ipv4_addresses": [
            "<ip-addr>"
        ],
       ...

```
Facts cache is not flushed.

**Test with --flush-cache:**
```
ansible servers -i ../ansible-test/inventory --flush-cache -m debug -a "msg={{ ansible_facts }}"

<output>
server1 | SUCCESS => {
    "msg": {}
}
```

Facts cache is flushed.

anvitpusalkar committed:
add --flush-cache option for the ansible command

anvitpusalkar commented:
Moved the `flush_cache` logic to CLI base class `_play_prereqs` method. Also, deleted the pattern form `playbook.py`.
Working on adding new tests for adhoc command.

anvitpusalkar commented:
Adding a test for `--flush-cache` in `test/integration/targets/adhoc` 

```
# --flush-cache option for ansible
# create facts_cache
rm -f localhost && ANSIBLE_CACHE_PLUGIN=jsonfile ANSIBLE_CACHE_PLUGIN_CONNECTION=. ansible localhost -m setup > /dev/null && test -s localhost
# flush facts_cache
ansible --flush-cache localhost -m debug -a "msg={{ ansible_facts }}" | grep '"msg": {}'
```
Suggestions?

s-hertel commented:
That looks good to me, just set the env vars in `test/integration/targets/adhoc/runme.sh` before the two tasks so the fact cache is also active for the test. Since ansible-test runs the tests in a tmp dir, there's no need to rm localhost.

```bash
# test ansible --flush-cache
export ANSIBLE_CACHE_PLUGIN=jsonfile
export ANSIBLE_CACHE_PLUGIN_CONNECTION=./
# collect and cache facts
ansible localhost -m setup > /dev/null && test -s localhost
# test flushing the fact cache
ansible --flush-cache localhost -m debug -a "msg={{ ansible_facts }}" | grep '"msg": {}'
```

anvitpusalkar committed:
add --flush-cache option for the ansible command

s-hertel commented:
@anvitpusalkar Thanks for the pull request!

anvitpusalkar commented:
Thanks for the review. Looking forward to further contributions!
